client/client.go

```diff
@@ -267,6 +269,80 @@ func vsockDial(host, port string) (net.Conn, string, error) {
 
 }
 
+func contains(elems []string, v string) bool {
+	for _, s := range elems {
+		if v == s {
+			return true
+		}
+	}
+	return false
+}
+
+// check that mdns response has all required attributes
+func mdns_required(src []string, req map[string][]string) bool {
+	txt := make(map[string]string)
+	for element := range src {
+		s := strings.SplitN(src[element], "=", 2)
+		txt[s[0]] = s[1]
+	}
+	for k, _ := range req {
+		if !contains(req[k], txt[k]) {
+			return false
+		}
+	}
+	return true
+}
+
+// lookup based on hostname, return resolved host, port, network, and error
+// uri currently supported dnssd://domain/_service._network/instance?reqkey=reqvalue
+// default for domain is local, first path element is _ncpu._tcp, and instance is wildcard
+// can omit to underspecify, e.g. dnssd:?arch=arm64 to pick any arm64 cpu server
+func mdns_lookup(uri string) (string, string, string, error) {
+	var (
+		hostname string
+		port     string
+		network  string
+		err      error
+	)
+
+	// determine if we want to do mdns
+	u, err := url.Parse(uri)
+	if (err == nil) && (u.Scheme == "dnssd") {
+		mdns_type := "_ncpu._tcp"
+		mdns_domain := "local"
+
+		// following dns-sd URI conventions from CUPS
+		if u.Host != "" {
+			mdns_domain = u.Host
+		}
+		if u.Path != "" {
+			mdns_type = u.Path
+		}
+
+		mdns_txt := u.Query()
+
+		// Make a channel for results and start listening
+		entriesCh := make(chan *mdns.ServiceEntry, 64)
+		// Setup sensible defaults"
+		// Start the lookup
+		err = mdns.Lookup(mdns_type+"."+mdns_domain, entriesCh)
+		if err != nil {
+			for entry := range entriesCh {
+				V("Got new mdns entry: %v\n", entry)
+				if mdns_required(entry.InfoFields, mdns_txt) {
+					hostname = entry.Addr.String()
+					port = strconv.Itoa(entry.Port)
+					network = "tcp"
```

Comment:
is `network` the right name for this variable? Should it be `protocol`?

Reply:
I was following the paradigm of the client options which callas it network -- I do realize that its unlikely we'd actually get a unix or a vsock response back from mdns...but those are more network than protocol.  It starts to get fuzzy.

Reply:
it's a go thing ... they use net and network when they kind of mean protocol. This is due to Plan 9 root IIUC.